File: 2D.py
import numpy as np
import tkinter as tk
from tkinter import ttk
import matplotlib.pyplot as plt
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import time

def topology(to_alvo):
    if to_alvo == "alvo_I":
        n = 2
        A = np.array([[[0], [0], [0], [0], [0], [0], [0]],
                      [[0], [1], [1], [1], [1], [1], [0]],
                      [[0], [0], [0], [1], [0], [0], [0]],
                      [[0], [0], [0], [1], [0], [0], [0]],
                      [[0], [0], [0], [1], [0], [0], [0]],
                      [[0], [1], [1], [1], [1], [1], [0]],
                      [[0], [0], [0], [0], [0], [0], [0]]])
    elif to_alvo == "alvo_U":
        n = 2
        A = np.array([[[0], [0], [0], [0], [0], [0], [0]],
                      [[0], [1], [0], [0], [0], [1], [0]],
                      [[0], [1], [0], [0], [0], [1], [0]],
                      [[0], [1], [0], [0], [0], [1], [0]],
                      [[0], [1], [0], [0], [0], [1], [0]],
                      [[0], [1], [1], [1], [1], [1], [0]],
                      [[0], [0], [0], [0], [0], [0], [0]]])
    elif to_alvo == "alvo_G":
        n = 2
        A = np.array([[[0], [0], [0], [0], [0], [0], [0]],
                      [[0], [1], [1], [1], [1], [1], [0]],
                      [[0], [1], [0], [0], [0], [0], [0]],
                      [[0], [1], [0], [1], [1], [1], [0]],
                      [[0], [1], [0], [0], [0], [1], [0]],
                      [[0], [1], [1], [1], [1], [1], [0]],
                      [[0], [0], [0], [0], [0], [0], [0]]])
    elif to_alvo == "alvo_dama":
        n = 2
        A = np.array([[[0], [1], [0], [1], [0], [1], [0]],
                      [[1], [0], [1], [0], [1], [0], [1]],
                      [[0], [1], [0], [1], [0], [1], [0]],
                      [[1], [0], [1], [0], [1], [0], [1]],
                      [[0], [1], [0], [1], [0], [1], [0]],
                      [[1], [0], [1], [0], [1], [0], [1]],
                      [[0], [1], [0], [1], [0], [1], [0]]])
    else:
        raise ValueError(f"Topo '{to_alvo}' não implementado.")
    return A, *A.shape, n

def plot_topology(ax, A, title="Topology", cycle=None):
    ax.clear()
    ax.imshow(A[:, :, 0], cmap="gray", origin="upper")
    ax.set_title(f"{title} (Cycle: {cycle})" if cycle else title)
    ax.axis("off")
    plt.draw()

def initialize_random_topology(shape, n):
    return np.random.randint(0, n, size=shape)

def aco_optimize(to_alvo, max_cycles=150, ax_goal=None, ax_process=None):
    A, nlA, ncA, pA, n = topology(to_alvo)
    m = 80  # Número de formigas
    tau = np.ones((n, nlA * ncA * pA)) / n

    best_topology = initialize_random_topology(A.shape, n)
    best_error = float("inf")

    plot_topology(ax_goal, A, title="Goal Topology")

    errors = []
    times = []  # Lista para armazenar tempos de execução acumulados
    start_time = time.time()

    for cycle in range(max_cycles):
        tabu = np.zeros((m, nlA, ncA, pA), dtype=int)
        
        for k in range(m):
            for x in range(nlA):
                for y in range(ncA):
                    for z in range(pA):
                        index = x * ncA * pA + y * pA + z
                        prob = tau[:, index]
                        prob_sum = np.sum(prob)
                        if prob_sum > 0:
                            prob /= prob_sum
                        else:
                            prob = np.ones(n) / n
                        prob = np.clip(prob, 0, 1)
                        prob /= prob.sum()
                        tabu[k, x, y, z] = np.random.choice(range(n), p=prob)

        fitness = np.array([np.sum(np.abs(tabu[k] - A)) for k in range(m)])
        min_fitness = fitness.min()
        if min_fitness < best_error:
            best_error = min_fitness
            best_topology = tabu[fitness.argmin()]

        tau *= 0.8
        for k in range(m):
            for x in range(nlA):
                for y in range(ncA):
                    for z in range(pA):
                        index = x * ncA * pA + y * pA + z
                        if fitness[k] > 0:  # Verificação para evitar divisão por zero
                            tau[tabu[k, x, y, z], index] += 1 / fitness[k]

        plot_topology(ax_process, best_topology, title="Processing Topology", cycle=cycle)
        plt.pause(0.1)

        errors.append(best_error)
        times.append(time.time() - start_time)

        if best_error == 0:
            print(f"Convergiu no ciclo {cycle}")
            break

    # Plota Ciclo x Erro
    plt.figure(figsize=(10, 5))
    plt.plot(errors, label="Error during execution - Cycle x Error")
    plt.xlabel("Cycles")
    plt.ylabel("Error")
    plt.title("Error Evolution - Cycle x Error")
    plt.grid(True)
    plt.legend()
    plt.show()

    # Plota Tempo x Erro
    plt.figure(figsize=(10, 5))
    plt.plot(times, errors, label="Error during execution - Time x Error", color="orange")
    plt.xlabel("Time (s)")
    plt.ylabel("Error")
    plt.title("Error Evolution - Time x Error")
    plt.grid(True)
    plt.legend()
    plt.show()

def main():
    def start_aco():
        to_alvo = selected_target.get()
        aco_optimize(to_alvo, ax_goal=ax_goal, ax_process=ax_process)

    root = tk.Tk()
    root.title("ACO Optimization - 2D")

    selected_target = tk.StringVar()
    selected_target.set("alvo_I")

    targets = ["alvo_I", "alvo_U", "alvo_G", "alvo_dama"]
    target_menu = ttk.Combobox(root, textvariable=selected_target, values=targets)
    target_menu.pack()

    start_button = tk.Button(root, text="Start Optimization", command=start_aco)
    start_button.pack()

    fig, (ax_goal, ax_process) = plt.subplots(1, 2, figsize=(10, 5))
    canvas = FigureCanvasTkAgg(fig, master=root)
    canvas.get_tk_widget().pack()

    root.mainloop()

if __name__ == "__main__":
    main()

File: 2D.c
import numpy as np import cupy as cp import matplotlib.pyplot as plt

    def initialize_pheromones(size, n) : ""
                                         "Inicializa a matriz de feromônios."
                                         "" return cp.ones((size, n)) / n

                                                           def construct_solutions(tabu, tau, nlA, ncA, pA, n) : ""
                                                                                                                 "Construção das soluções pelas formigas."
                                                                                                                 "" num_ants = tabu.shape[0]

                                                                                   for ant in range(num_ants) : for x in range(nlA) : for y in range(ncA) : for z in range(pA) :index = x * ncA * pA + y * pA + z prob = tau[index]

#Normaliza probabilidades
                                                                                                                                                                                                                             prob_sum = cp.sum(prob) if prob_sum> 0 :prob /= prob_sum else :prob = cp.ones(n) / n

#Escolha do material
                                                                                                                                                                                                                                                                                                           choice = cp.random.choice(n, p = cp.asnumpy(prob)) tabu[ant, x, y, z] = choice

                                                                                                                                                                                                                                                                                                                                                                   def update_pheromones(tau, tabu, fitness, nlA, ncA, pA, n) : ""
                                                                                                                                                                                                                                                                                                                                                                                                                                "Atualização dos feromônios."
                                                                                                                                                                                                                                                                                                                                                                                                                                "" tau *= 0.8 #Evaporação num_ants = tabu.shape[0]

                                                                                                                                                                                                                                                                                                                                                                                         for ant in range(num_ants) : for x in range(nlA) : for y in range(ncA) : for z in range(pA) :index = x * ncA * pA + y * pA + z material = tabu[ant, x, y, z] tau[index, material] += 1.0 / fitness[ant]

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          def evaluate_solutions(tabu, goal, nlA, ncA, pA) : ""
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             "Avaliação das soluções."
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             "" num_ants = tabu.shape[0] fitness = cp.zeros(num_ants)

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          for ant in range(num_ants) :fitness[ant] = cp.sum(cp.abs(tabu[ant] - goal))

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                return fitness

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            def main() :nlA, ncA, pA = 7, 7, 1 #Dimensões do espaço de design n = 2 #Número de materiais possíveis num_ants = 50 max_cycles = 150

#Inicialização
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     size = nlA * ncA * pA tau = initialize_pheromones(size, n) tabu = cp.zeros((num_ants, nlA, ncA, pA), dtype = int) goal = cp.random.randint(0, n, size =(nlA, ncA, pA))

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          best_fitness = float("inf") best_solution = None

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                fitness_history =[]

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                for cycle in range(max_cycles) :
#Construção de soluções
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                construct_solutions(tabu, tau, nlA, ncA, pA, n)

#Avaliação
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        fitness = evaluate_solutions(tabu, goal, nlA, ncA, pA) min_fitness = cp.min(fitness) best_ant = cp.argmin(fitness)

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        if min_fitness < best_fitness:
            best_fitness = min_fitness
            best_solution = tabu[best_ant]

        fitness_history.append(best_fitness)

#Atualização dos feromônios
        update_pheromones(tau, tabu, fitness, nlA, ncA, pA, n)

        print(f"Cycle {cycle}: Best Fitness = {best_fitness}")

        if best_fitness == 0:
            print(f"Converged at cycle {cycle}")
            break

#Visualização dos resultados
    plt.plot(fitness_history, label="Fitness over cycles")
    plt.xlabel("Cycles")
    plt.ylabel("Fitness")
    plt.title("ACO Optimization - Fitness Evolution")
    plt.legend()
    plt.show()

if __name__ == "__main__":
    main()
